main.go

```diff
@@ -18,96 +17,68 @@ var (
 	verbose     = false
 	dryrun      = false
 	showVersion = false
-
-	commandName      = "/sbin/zfs"
-	commandArguments = []string{"list", "-t", "snapshot", "-o", "name,creation", "-s", "creation", "-r", "-H", "-p"}
-
 	// This can be set to a specific time for testing.
 	now = time.Now()
-
 	// tasks can be added to this for testing.
 	mainWaitGroup sync.WaitGroup
-
 	// This can be changed to true when testing.
 	panicBail = false
-
-	rootCmd = &cobra.Command{
+	rootCmd   = &cobra.Command{
 		Use:   "zfs-cleaner [config file]",
 		Short: "Tool for destroying ZFS snapshots after predefined retention periods",
 		RunE:  clean,
 	}
-
 	// Can be overridden when running tests.
-	stdout io.Writer = os.Stdout
+	stdout      io.Writer = os.Stdout
+	zfsExecutor zfs.Executor
 )
 
 func init() {
 	rootCmd.PersistentFlags().BoolVarP(&dryrun, "dryrun", "n", false, "Do nothing destructive, only print")
 	rootCmd.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, "Be more verbose")
 	rootCmd.PersistentFlags().BoolVarP(&showVersion, "version", "V", false, "Show version and exit")
 	rootCmd.TraverseChildren = true
+	zfsExecutor = zfs.NewExecutor()
 }
 
-func getList() ([]byte, error) {
-	output, err := exec.Command(commandName, commandArguments...).Output()
-
-	if err != nil {
-		return nil, err
-	}
-
-	return output, nil
-}
-
-func readConf(r *os.File) (*conf.Config, error) {
-	conf := &conf.Config{}
-
-	err := conf.Read(r)
+func readConfig(r *os.File) (*conf.Config, error) {
+	config := &conf.Config{}
+	err := config.Read(r)
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse %s: %s", r.Name(), err.Error())
 	}
-
-	return conf, nil
+	return config, nil
 }
 
-func processAll(now time.Time, conf *conf.Config) ([]zfs.SnapshotList, error) {
-	list, err := getList()
-	if err != nil {
-		return nil, err
-	}
-
+func processAll(now time.Time, conf *conf.Config, zfsExecutor zfs.Executor) ([]zfs.SnapshotList, error) {
 	lists := []zfs.SnapshotList{}
-
 	for _, plan := range conf.Plans {
-		for _, path := range plan.Paths {
-			list, err := zfs.NewSnapshotListFromOutput(list, path)
+		for _, dataset := range plan.Paths {
+			list := zfs.SnapshotList{}
+			list, err := list.NewSnapshotListFromDataset(zfsExecutor, dataset)
 			if err != nil {
 				return nil, err
 			}
-
 			list.KeepNamed(plan.Protect)
 			list.KeepLatest(plan.Latest)
-
 			for _, period := range plan.Periods {
 				start := now.Add(-period.Age)
-
 				list.Sieve(start, period.Frequency)
 			}
-
 			lists = append(lists, list)
 		}
 	}
-
 	return lists, nil
 }
 
 func main() {
+	AddPlanCheckCommand(zfsExecutor)
 	err := rootCmd.Execute()
 	if err != nil {
 		if panicBail {
 			panic(err.Error())
 		}
-
-		fmt.Fprintf(os.Stderr, "%s\n", err.Error())
+		_, _ = fmt.Fprintf(os.Stderr, "%s\n", err.Error())
```

Comment:
if we ignore both return result i would think that simply omitting the `_` throwaway assignments, especially on a `fmt` call, is more readable

Reply:
This is more explicit